Begin PAL'ing out S.N.Sockets.

pgavlin opened the pull request:
- PAL out all synchronous methods for Sockets
  - `ioctl` is an open issue
- PAL out the APM methods for Sockets
- Implement the synchronous and APM PALs for Linux

Some things are still a bit messy: e.g. there are some shims in place to deal with the fact that we're still building some of the Windows bits on Linux.

pgavlin committed:
Pal out synchronous methods for S.N.Sockets.

pgavlin committed:
PAL out asynchronous methods for S.N.Sockets.

pgavlin committed:
Move to EPOLLET; add SendTo/ReceiveFrom

pgavlin committed:
Move some PAL bits around.

pgavlin committed:
Fix issues with Windows.

pgavlin commented:
@CIPop PTAL

pgavlin committed:
Split various types out of Socket.cs.

This change moves SocketAsyncEventArgs, SocketAsyncOperation, and
SendPacketsElement into their own files.

pgavlin committed:
PAL out SocketAsyncEventArgs and Socket.SendMultiple.

pgavlin committed:
PAL out the Socket.*Async APIs.

pgavlin committed:
Initial cut at SocketAsyncEventArgs for *nix.

pgavlin committed:
SocketAsyncContext: do not call completion callbacks synchronously.

This more closely matches the behavior of I/O completion ports on Windows.

pgavlin commented:
FYI: I've pushed new commits that add support for most of the `Socket.*Async` APIs.

CIPop commented on src/Common/src/System/Net/SafeCloseSocket.Windows.cs:
```diff
@@ -0,0 +1,244 @@
+// Copyright (c) Microsoft. All rights reserved.
```
For some reason SafeCloseSocket isn't shown as copied. Were there any changes made to this from the original?

CIPop commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs:
```diff
@@ -0,0 +1,1233 @@
+        {
+            get
+            {
+                return m_SendPacketsDescriptor == null ? null : m_sendPacketsDescriptor.Length;
```
Fails compilation: needs capital s. I think you meant 

`return m_SendPacketsDescriptor == null ? null : (int?)m_SendPacketsDescriptor.Length;`

Severity    Code    Description Project File    Line    Source  Tool
Error   CS0103  The name 'm_sendPacketsDescriptor' does not exist in the current context    System.Net.Sockets  S:\c1\src\System.Net.Sockets\src\System\Net\Sockets\SocketAsyncEventArgs.Windows.cs 87  IntelliSense    Live

pgavlin replied:
Yes, good catch. I forgot to try building the *Async PAL on Windows.

CIPop commented on src/Native/System.Native/pal_io.cpp:
```diff
@@ -359,6 +359,39 @@ int32_t FcntlSetPipeSz(int32_t fd, int32_t size)
 }
```
Shouldn't this native code be split out between OSs? This seems applicable to only *NIX/OSX.

pgavlin replied:
This is only used on *nix and OS X :)

CIPop replied:
Sorry I wasn't very clear... What I meant was maybe this could be placed in a folder like System.Native.Unix instead of System.Native. This way the intent is very clear.

pgavlin replied:
You'll have to take that up with Nick Guerrera.

CIPop replied:
Taking this offline.

CIPop commented on src/System.Net.Sockets/src/RefactoringShims/InteropShim.cs:
```diff
@@ -0,0 +1,62 @@
+// Copyright (c) Microsoft. All rights reserved.
```
I can't find references to this in the Linux_Debug configuration. Is this still needed?

pgavlin replied:
Yes. It's included under `<ItemGroup Condition=" '$(TargetsUnix)' == 'true' ">`, which triggers for Linux, OS X, and the various BSDs.

CIPop replied:
VS didn't show any reference... it's probably caused by the bug that switched "/" with "\" within dir.targets.
I tried removing it and now I understand what's going on: you are still including some of the classes I mistakenly placed in Interop\Winsock. 

I think we'll need to PAL these as well and move them to common.

CIPop commented on src/System.Net.Sockets/src/System.Net.Sockets.csproj:
```diff
@@ -223,9 +213,178 @@
+    <Compile Include="System\Net\Sockets\_BaseOverlappedAsyncResult.Unix.cs" />
+    <Compile Include="System\Net\Sockets\_ConnectOverlappedAsyncResult.Unix.cs" />
+    <Compile Include="System\Net\Sockets\_OverlappedAsyncResult.Unix.cs" />
+    <Compile Include="System\Net\Sockets\BufferList.cs" />
```
BufferList is a public API in System.Net.Sockets 4.0.0.0 (already released). Please move this to the common ItemGroup.

pgavlin replied:
This (perhaps confusingly named) file contains an internal type that I added for Unix/Linux, nothing else. It does not relate to any existing APIs.

CIPop replied:
I see it now. Then it needs to:
- be clearly marked as *Nix only (maybe BufferList.Unix.cs ?)
- maybe have a different name to avoid confusion with `SocketAsyncEventArgs.BufferList`

pgavlin replied:
Would you be opposed to me keeping the type name (it's concise and descriptive) and just moving the definition of the type into SocketAsyncContext.cs (which is where its primary consumer lives)?

CIPop replied:
Sounds good.

CIPop commented on src/System.Net.Sockets/src/System.Net.Sockets.csproj:
```diff
@@ -223,9 +213,178 @@
+    <Compile Include="System\Net\Sockets\SocketPal.Unix.cs" />
+
+    <!-- TODO: remove this -->
+    <Compile Include="RefactoringShims\InteropShim.cs">
```
See above: no references to the types. Is it really needed?

pgavlin replied:
The references come from the Windows interop files that are still included when building for *nix.

CIPop replied:
Saw the references now.

CIPop commented on src/System.Net.Sockets/src/System.Net.Sockets.csproj:
```diff
@@ -223,9 +213,178 @@
+      <Link>Interop\Linux\libc\Interop.epoll_wait.cs</Link>
+    </Compile>
+
+    <Compile Include="$(CommonPath)\Interop\Unix\libc\Interop.addrinfo.cs">
```
These files are marked as "Unix". Why not have the same conditional on the ItemGroup as in System.Net.Primitives.csproj. (Same question applies for NameResolution.)

pgavlin replied:
Frankly, these files should just be moved into Interop\Linux.

CIPop commented on src/System.Net.Sockets/src/System.Net.Sockets.csproj:
```diff
@@ -223,9 +213,178 @@
+
+  <ItemGroup Condition=" '$(TargetsLinux)' == 'true' ">
+    <Compile Include="System\Net\Sockets\SocketAsyncContext.cs" />
+    <Compile Include="System\Net\Sockets\SocketAsyncEngine.cs" />
```
These two files need to be marked as .Linux I presume? Or do you plan to have these part of the PAL layer?

pgavlin replied:
They can  be marked as Linux, yes. The picture will get a bit more complicated when OS X enters the picture, however: I expect to reuse large parts of these two types (only the event notification mechanism should be different between the two OSes), so these will probably become SocketAsync{Context,Engine}.{Unix,Linux,OSX}.cs over time.

CIPop commented on src/System.Net.Sockets/src/System/Net/Sockets/BufferList.cs:
```diff
@@ -0,0 +1,55 @@
+using System.Collections;
+using System.Collections.Generic;
+
+namespace System.Net.Sockets
```
In case this file changed: please ensure that the public API supported in the 4.0.0.0 and 4.0.10.0 contracts is still available.

pgavlin replied:
As mentioned, this file is new: it contains an internal type I used for the Linux async engine to support polymorphism over `IList<ArraySegment<byte>>` and `BufferOffsetSize[]`.

CIPop replied:
Saw the above comment: thanks for explaining.

As above: maybe renaming the inner type as well as ensuring that the file-name ends in .Linux.cs would avoid later confusion.

CIPop replied:
Comment resolved by above discussion.

CIPop commented on src/System.Net.Sockets/src/System.Net.Sockets.csproj:
```diff
@@ -223,9 +213,178 @@
+    <Compile Include="System\Net\Sockets\_ConnectOverlappedAsyncResult.Unix.cs" />
+    <Compile Include="System\Net\Sockets\_OverlappedAsyncResult.Unix.cs" />
+    <Compile Include="System\Net\Sockets\BufferList.cs" />
+    <Compile Include="System\Net\Sockets\Socket.Unix.cs" />
```
Shouldn't this be part of SocketPal?

pgavlin replied:
Same answer as for `SocketAsyncEventArgs`.

CIPop commented:
Can't review src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs - GitHub is saying:
125 additions, 2,769 deletions not shown

Comment on Socket.cs: Shouldn't Socket.Windows/Unix be part of SocketPal?

CIPop commented on src/System.Net.Sockets/src/System.Net.Sockets.csproj:
```diff
@@ -223,9 +213,178 @@
+    <Compile Include="System\Net\Sockets\_BaseOverlappedAsyncResult.Windows.cs" />
+    <Compile Include="System\Net\Sockets\_ConnectOverlappedAsyncResult.Windows.cs" />
+    <Compile Include="System\Net\Sockets\_OverlappedAsyncResult.Windows.cs" />
+    <Compile Include="System\Net\Sockets\Socket.Windows.cs" />
```
Shouldn't this be part of SocketPal?

pgavlin replied:
Same answer as for `SocketAsyncEventArgs`.

CIPop commented on src/System.Net.Sockets/src/System.Net.Sockets.csproj:
```diff
@@ -223,9 +213,178 @@
+    <Compile Include="System\Net\Sockets\_ConnectOverlappedAsyncResult.Windows.cs" />
+    <Compile Include="System\Net\Sockets\_OverlappedAsyncResult.Windows.cs" />
+    <Compile Include="System\Net\Sockets\Socket.Windows.cs" />
+    <Compile Include="System\Net\Sockets\SocketAsyncEventArgs.Windows.cs" />
```
Shouldn't a SocketAsyncEventArgsPal exist instead?

pgavlin replied:
That's effectively what SocketAsyncEventArgs.{Windows,Unix}.cs are: all of the platform-specific implementation details for `SocketAsyncEventArgs` are within these files. The implementation details need to access enough of the internal state of the type that it felt more efficient to lay the PAL out this way (i.e. as part of the type itself).

CIPop replied:
It's confusing to have two mechanisms for PAL.
See below comment regarding public / PAL interfaces.

CIPop commented:
Socket.cs:5361 AcceptEx is marked as internal. This creates a circular dependency between Socket (common xplat) and SocketPal. 
A better design would limit the dependency within SocketPal. This probably means that Socket::AcceptEx should be moved within SocketPal.

Same with DisconnectEx just that it takes a dependency on the particular implementation for SocketAsyncEventArgs.Windows.cs.

CIPop commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.cs:
```diff
@@ -0,0 +1,1197 @@
+namespace System.Net.Sockets
+{
+    // TODO
+    // - ReceiveMessageFrom
```
Also, please add a TODO to move this to Linux/Unix specific code.

pgavlin replied:
Not sure what you mean--this is already Linux/Unix specific. Are you just referring to changing the file name?

CIPop replied:
Yes, just the file name.

pgavlin commented:
> Socket.cs:5361 AcceptEx is marked as internal. This creates a circular dependency between Socket (common xplat) and SocketPal. 
> A better design would limit the dependency within SocketPal. This probably means that Socket::AcceptEx should be moved within SocketPal.
> 
> Same with DisconnectEx just that it takes a dependency on the particular implementation for SocketAsyncEventArgs.Windows.cs.

Yes, I agree. I did not want to move this code around quite yet, but I'm happy to do so now if you think that it's OK.

CIPop commented:
> > Socket.cs:5361 AcceptEx is marked as internal. This creates a circular dependency between Socket (common xplat) and SocketPal. 
> > A better design would limit the dependency within SocketPal. This probably means that Socket::AcceptEx should be moved within SocketPal.
> > 
> > Same with DisconnectEx just that it takes a dependency on the particular implementation for SocketAsyncEventArgs.Windows.cs.
> 
> Yes, I agree. I did not want to move this code around quite yet, but I'm happy to do so now if you think that it's OK.

I think we should clearly separate the public interfaces from inner implementation details. 
For that, we could decide on always creating a separate class which holds most of the state and separate PAL partial class implementations for various OS flavors:
- Socket.cs (public interface, has no or little state) uses
  - SocketPAL.cs (common PAL layer, holds most state, no references to public Socket.cs APIs)
    - SocketPAL.Windows.cs (specific OS implementation)
    - SocketPAL.Linux.cs
    - SocketPAL.OSX.cs
- SocketAsyncEvent.cs (public interface, Socket.cs can reference this but none of the PAL classes can reference back to it)
  - SocketAsyncEventPAL.cs (common PAL layer) can be referenced by SocketPAL.cs directly
    - SocketAsyncEventPAL.Windows.cs
    - etc

pgavlin commented:
> I think we should clearly separate the public interfaces from inner implementation details.

I'm not sure how that's not achieved here. The public interfaces are in Socket.cs, SocketAsyncEventArgs.cs, etc. The OS-specific files contain OS-specific implementation details only.

> Socket.cs (public interface, has no or little state) uses

Unfortunately, "has no or little state" is not achievable: `Socket` and its related types have a fair bit of state that is relevant to the proper operation of the APIs no matter what platform we're running on.

CIPop commented:
> I'm not sure how that's not achieved here. The public interfaces are in Socket.cs, SocketAsyncEventArgs.cs, etc. The OS-specific files contain OS-specific implementation details only.

It is achieved but in two different ways: one via the SocketPAL class and one via having Socket itself be a partial class. 

> Unfortunately, "has no or little state" is not achievable:  Socket  and its related types have a fair bit of state that is relevant to the proper operation of the APIs no matter what platform we're running on.

Why not keep Socket.cs as a public interface (we could have no state at all, just a simple pass-through to the SocketPAL class).
This means copy/pasting everything in SocketPAL and leaving Socket.cs to simply call into SocketPAL for each function. This interface separation is a recurring pattern: we're using it when designing COM/WRL classes.
The only downside is that Socket.cs will have an additional function call for each API after validating user input. The perf impact is really nothing compared to the benefit of code clarity during later support (and in most modern compilers these calls should actually be optimized in-line anyway).

Example:

```
Socket.cs:
       public void Connect(EndPoint remoteEP)
       {
            if (remoteEP == null)
            {
                throw new ArgumentNullException("remoteEP");
            }

            SocketPal.Connect(remoteEP);
        }          
```

An example for a property that already does this is `Socket::OSSupportsIPv4`. We should move `InitializeSockets()` to SocketPal and leave the implementation as:

```
        public static bool OSSupportsIPv4
        {
            get
            {
                return SocketPal.OSSupportsIPv4;
            }
        }
```

pgavlin commented:
> This means copy/pasting everything in SocketPAL and leaving Socket.cs to simply call into SocketPAL for each function. This interface separation is a recurring pattern: we're using it when designing COM/WRL classes.

If I understand correctly, what you're suggesting involves duplicating any code that touches internal state, even if that state is shared between platforms. I don't think that this pattern is a good fit for Sockets: there is a sizable amount of state and behavior that can and should be shared between the various platforms in order to better ensure that they produce mutually compatible behavior.

In general, I see a number of options in this scenario (i.e. a type that has non-trivial amounts of both platform-independent and platform-dependent state and behavior):
1. Use a partial class to share platform-independent code with platform-specific code. This option is simple and expedient, but may over-expose the platform-independent details to the platform-specific code. In this case, I felt that the simplicity and expediency outweighed the downsides.
2. Push the platform-specific state into derived types. This option does not apply to sockets b/c the types involved are public and unsealed.
3. Push the platform-specific state into new value types that expose an internal API. The platform-independent types then keep a value of the platform-specific type in a field and expose state as necessary via the internal API. This option is more complicated than (1), but has the benefit of preventing over-exposure of platform-independent details.
4. Expose the platform-independent state as internal APIs/fields and push the platform-specific state into new types that consume these fields. This option is of similar complexity to (3), but grossly over-exposes the platform-independent details.

Long-term, I think that the most palatable option is probably (3). It should be fairly straightforward to migrate to such a design from where we are now.

FWIW, the reason that we haven't yet hit these design issues is that the other PALs that I've added are for the most part stateless, and as such can be implemented as a static class that exports an internal API.

pgavlin commented:
@stephentoub would you mind taking a look at this as well?

pgavlin committed:
Address CR feedback.

- Rename some files to indicate that they are Linux-specific
- Fix a break in the Windows build.

CIPop commented:
Thanks for explaining!

The reasoning behind by initial comment is a [code smell](https://en.wikipedia.org/wiki/Code_smell): In my mind, if it's not marked as PAL it shouldn't be a partial class and shouldn't have .Linux and .Windows versions. (It's the _common_ part.)

> Long-term, I think that the most palatable option is probably (3). It should be fairly straightforward to migrate to such a design from where we are now.

I agree that this is the right solution. (Sounds like the [strategy pattern](https://en.wikipedia.org/wiki/Strategy_pattern).) My solution would have only solved my concern regarding code organization but not the real issue which is better separation of internal state (context) used by different OS-specific implementation (strategies).

Do you think we could achieve this implementation now rather than after it got pulled in corefx?

pgavlin commented:
> Do you think we could achieve this implementation now rather than after it got pulled in corefx?

We can probably get to this before merging into corefx/master, yes. I'm going to leave it as-is for this PR, however.

pgavlin commented:
@CIPop should I just go ahead and merge this to unblock #12 and file an issue to follow-up on the CR process?

pgavlin committed:
Clarify some comments and fix the starting state of SocketAsyncContext's queues.

CIPop commented:
Yes, I think it's easier to get this code in as-is for now and have a smaller PR for the items discussed here.

stephentoub commented on src/Common/src/Interop/Unix/System.Native/Interop.Fcntl.cs:
```diff
@@ -20,6 +20,12 @@ internal static class Fcntl
+            internal static extern int GetIsNonBlocking(int fd);
+
+            [DllImport(Libraries.SystemNative, EntryPoint="FcntlSetIsNonBlocking", SetLastError=true)]
+            internal static extern int SetIsNonBlocking(int fd, int isNonBlocking);
```
Should we declare isNonBlocking as a bool with [MarshalAs(UnmanagedType.Bool)]?  It'd make call sites a bit clearer I think.  Same could be done for the return type of GetIsNonBlocking.  @nguerrera, any concerns with something like that?

stephentoub commented on src/Common/src/Interop/Unix/libc/Interop.SocketOptionName.cs:
```diff
@@ -0,0 +1,46 @@
+
+internal static partial class Interop
+{
+    // NOTE: this is only correct for Linux on x86-64
```
@nguerrera, just FYI.  Lots of stuff here ripe for the native shim picking :)

stephentoub commented on src/Common/src/Interop/Unix/libc/Interop.bind.cs:
```diff
@@ -0,0 +1,17 @@
+    internal static partial class libc
+    {
+        [DllImport(Libraries.Libc, SetLastError = true)]
+		public static extern unsafe int bind(int sockfd, sockaddr* addr, socklen_t addrlen);
```
Nit: spacing looks off

stephentoub commented on src/Common/src/Interop/Unix/libc/Interop.listen.cs:
```diff
@@ -0,0 +1,15 @@
+    internal static partial class libc
+    {
+        [DllImport(Libraries.Libc, SetLastError = true)]
+		public static extern int listen(int sockfd, int backlog);
```
Nit: spacing

stephentoub commented on src/Common/src/System/Net/SafeCloseSocket.Unix.cs:
```diff
@@ -0,0 +1,213 @@
+    internal sealed partial class SafeCloseSocket : DebugSafeHandleMinusOneIsInvalid
+#else
+    internal sealed partial class SafeCloseSocket : SafeHandleMinusOneIsInvalid
+#endif
```
To further highlight the debug difference, this could be:

``` C#
internal sealed partial class SafeCloseSocket :
#if DEBUG
    DebugSafeHandleMinusOneIsInvalid
#else
    SafeHandleMinusOneIsInvalid
#endif
```

stephentoub commented on src/Common/src/System/Net/SafeCloseSocket.Unix.cs:
```diff
@@ -0,0 +1,213 @@
+                {
+                    if (Volatile.Read(ref _asyncContext) == null)
+                    {
+                        Interlocked.CompareExchange(ref _asyncContext, new SocketAsyncContext((int)handle, SocketAsyncEngine.Instance), null);
```
Is there any cleanup that needs to be done on the allocated SocketAsyncContext if we lose the race?

pgavlin replied:
There is not, no (just the opposite, in fact: the losers must not remove the socket fd from the epoll instance).

stephentoub commented on src/Common/src/System/Net/SafeCloseSocket.Unix.cs:
```diff
@@ -0,0 +1,213 @@
+					if (errorCode == -1)
+					{
+						errorCode = (int)Interop.Sys.GetLastError();
+					}
```
Nit: spacing of the above block looks wrong (maybe tabs were used?)

stephentoub commented on src/Common/src/System/Net/SafeCloseSocket.Unix.cs:
```diff
@@ -0,0 +1,213 @@
+                        {
+                            _asyncContext.Close();
+                        }
+                        return SocketPal.GetSocketErrorForErrorCode((Interop.Error)errorCode);
```
This whole if block looks identical to the earlier code... consider refactoring it out into a helper method.

stephentoub commented on src/Common/src/System/Net/SafeCloseSocket.Unix.cs:
```diff
@@ -0,0 +1,213 @@
+                {
+                    _asyncContext.Close();
+                }
+                return SocketPal.GetSocketErrorForErrorCode((Interop.Error)errorCode);
```
Looks like another place the aforementioned helper could be used.

stephentoub commented on src/Common/src/System/Net/SafeCloseSocket.Unix.cs:
```diff
@@ -0,0 +1,213 @@
+                int sock = SocketPal.GetPlatformSocketType(socketType);
+                int pt = (int)protocolType;
+
+				int fd = Interop.libc.socket(af, sock, pt);
```
Nit: spacing

stephentoub commented on src/Common/src/System/Net/SafeCloseSocket.Windows.cs:
```diff
@@ -0,0 +1,244 @@
+                        {
+                            if (ExceptionCheck.IsFatal(exception)) throw;
+                            CloseAsIs();
+                            throw;
```
Might be a bit simpler as:

```
if (!ExceptionCheck.IsFatal(exception)) CloseAsIs();
throw;
```

stephentoub replied:
(I just realized this code is just being moved... I'll skip over these Windows files)

stephentoub commented on src/Native/System.Native/pal_io.cpp:
```diff
@@ -359,6 +359,39 @@ int32_t FcntlSetPipeSz(int32_t fd, int32_t size)
+    else
+    {
+        flags |= O_NONBLOCK;
+    }
```
How common will it be for this call to effectively be a nop, e.g. calling to set isNonBlocking==true when it's already non-blocking, or isNonBlocking==false when it's already blocking?  If it'll be common, we could avoid the extra sys call but comparing where the new flags are the same as the old, and skipping the fcntl call if they are.  If it won't be common, this looks good.

pgavlin replied:
It should not be common--each socket tracks its blocking mode--but the change is straightforward. Fixed.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SendPacketsElement.cs:
```diff
@@ -0,0 +1,124 @@
+        internal byte[] m_Buffer;
+        internal int m_Offset;
+        internal int m_Count;
+        internal Interop.Winsock.TransmitPacketsElementFlags m_Flags;
```
This file looks like it's used for both Windows and Unix, but it has some Winsock-specific logic?

pgavlin replied:
There is still some winsock-related code that is compiled for Unix; this code will be PAL'd out soon.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/Socket.Unix.cs:
```diff
@@ -0,0 +1,84 @@
+{
+    public partial class Socket
+    {
+        internal static int FillFdSetFromSocketList(ref Interop.libc.fd_set fdset, IList socketList)
```
I realize this would affect upstream code as well, but could the argument be changed to be an `IList<Socket>` instead?

pgavlin replied:
I don't think so: the input `IList` is provided via a public API.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Linux.cs:
```diff
@@ -0,0 +1,1200 @@
+    //     - This will require a new state for each queue, unregistred, to track whether or
+    //       not the queue is currently registered to receive events
+
+    sealed class SocketAsyncContext
```
Nit: we try to be explicit about visibility, e.g. internal

pgavlin replied:
Done.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Linux.cs:
```diff
@@ -0,0 +1,1200 @@
+        private abstract class AcceptOrConnectOperation : AsyncOperation
+        {
+            public byte[] SocketAddress;
+            public int SocketAddressLen;
```
I see that these two fields also exist in TransferOperation.  Any value in having TransferOperation derive from this type (with a different name)?  Or maybe these should just be pushed down to AsyncOperation?

pgavlin replied:
These fields can be pushed down.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Linux.cs:
```diff
@@ -0,0 +1,1200 @@
+            Clear = 1,
+        }
+
+        struct Queue<TOperation>
```
Consider using a different name, just so that folks reading the code don't confuse it with `System.Collections.Generic.Queue<T>`

pgavlin replied:
Renamed to `OperationQueue<TOperation>`.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Linux.cs:
```diff
@@ -0,0 +1,1200 @@
+                    _tail.Next = operation;
+                }
+
+                _tail = operation;
```
Maybe assert that operation.Next == operation?

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Linux.cs:
```diff
@@ -0,0 +1,1200 @@
+
+            public void Enqueue(TOperation operation)
+            {
+                Debug.Assert(!IsStopped);
```
Maybe assert that operation.Next == operation?

pgavlin replied:
Good idea. Done.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Linux.cs:
```diff
@@ -0,0 +1,1200 @@
+        object _closeLock = new object();
+        object _queueLock = new object();
+        SocketAsyncEngine _engine;
+        uint _registeredEvents;
```
Nit: we should be explicit about the private visibility on all of these

pgavlin replied:
Done.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Linux.cs:
```diff
@@ -0,0 +1,1200 @@
+            Interop.Error errorCode;
+            if (!_engine.TryRegister(this, _fileDescriptor, events, 0, ref _handle, out errorCode))
+            {
+                throw new Exception(string.Format("SocketAsyncContext: {0}", errorCode));
```
Presumably we'll go back and update both the type of these exceptions and their error messages?  Maybe add a TODO in the code somewhere to follow-up on that?

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Linux.cs:
```diff
@@ -0,0 +1,1200 @@
+        public void Close()
+        {
+            lock (_closeLock)
+            lock (_queueLock)
```
It'd be worth adding a comment where these locks are defined about the lock hierarchy that needs to be followed to avoid deadlocks.  And then in places where _closeLock is acquired, asserting that _queueLock isn't held (`Debug.Assert(!Monitor.IsEntered(_queueLock));`).

pgavlin replied:
Done.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Linux.cs:
```diff
@@ -0,0 +1,1200 @@
+                        return false;
+
+                    case State.Clear:
+                        break;
```
You have this here just to be explicit about all of the cases?  If so, consider instead doing:

``` C#
default:
    Debug.Assert(queue.State == State.Clear);
    break;
```

pgavlin replied:
I have this here to be explicit about the actions taken at each state. Personally, I'd prefer to keep this as-is.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Linux.cs:
```diff
@@ -0,0 +1,1200 @@
+            SocketError errorCode;
+            if (TryCompleteAccept(_fileDescriptor, socketAddress, ref socketAddressLen, out acceptedFd, out errorCode))
+            {
+                ThreadPool.QueueUserWorkItem(_ => callback(acceptedFd, socketAddress, socketAddressLen, errorCode), null);
```
As written, this is going to cause a closure allocation even if TryCompleteAccept returns false, as it's capturing state outside of the scope of the if block.  It's also going to cause a delegate allocation when invoked.  You could address both of those by changing it to something like:

```
ThreadPool.QueueUserWorkItem(s =>
{
    var args = (Tuple<...>)s;
    args.Item1(args.Item2, args.Item3, args.Item4, args.Item5);
}, Tuple.Create(callback, acceptedFd, socketAddress, socketAddressLen, errorCode));
```

pgavlin replied:
Done.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Linux.cs:
```diff
@@ -0,0 +1,1200 @@
+            SocketError errorCode;
+            if (TryCompleteConnect(_fileDescriptor, socketAddress, socketAddressLen, out errorCode))
+            {
+                ThreadPool.QueueUserWorkItem(_ => callback(errorCode), null);
```
I won't comment on these closures further; at some point it'd be good to do a pass through to fix up such low-hanging fruit.  But I think you said many of these synchronously completing cases are likely to go away anyway, such that we don't invoke the callback asynchronously (great!), in which case the problem will mostly solve itself.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Linux.cs:
```diff
@@ -0,0 +1,1200 @@
+            {
+                if (socketAddress != null)
+                {
+                    pinnedSocketAddress = GCHandle.Alloc(socketAddress, GCHandleType.Pinned);
```
Why this rather than using fixed?

pgavlin replied:
Does `fixed` work when the object being pinned is `null`? The C# spec leaves the behavior up to the implementation (see section 18.6 of the C# 5.0 spec: "The behavior of the fixed statement is implementation-defined if the array expression is null or if the array has zero elements.").

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Linux.cs:
```diff
@@ -0,0 +1,1200 @@
+                    pinnedSocketAddress = GCHandle.Alloc(socketAddress, GCHandleType.Pinned);
+                    sockAddr = (Interop.libc.sockaddr*)pinnedSocketAddress.AddrOfPinnedObject();
+                    sockAddrLen = (uint)socketAddressLen;
+                }
```
Same question as before about why GCHandle.Alloc instead of fixed.  I'm probably missing something obvious?

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Linux.cs:
```diff
@@ -0,0 +1,1200 @@
+                {
+                    ArraySegment<byte> buffer = buffers[startIndex + i];
+                    handles[i] = GCHandle.Alloc(buffer.Array, GCHandleType.Pinned);
+                    iovecs[i].iov_base = &((byte*)handles[i].AddrOfPinnedObject())[buffer.Offset + startOffset];
```
Is it possibl for the caller-provided offset to extend beyond the length of the first buffer?

pgavlin replied:
No. I'll add an assert to that effect.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Linux.cs:
```diff
@@ -0,0 +1,1200 @@
+                                {
+                                    // TODO: throw an appropiate exception
+                                    throw new Exception(string.Format("HandleEvents HUP: {0}", errorCode));
+                                }
```
We don't want to free this handle in the case of an exception?  Is that because the caller might try to recover and continue using it?

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Linux.cs:
```diff
@@ -0,0 +1,1200 @@
+                        else
+                        {
+                            completed = TryCompleteConnect(_fileDescriptor, (ConnectOperation)op);
+                        }
```
I think I've heard you say you don't like this construction, but FWIW, I prefer seeing such statements as:

``` C#
bool completed = (acceptOp != null) ?
    TryCompleteAccept(_fileDescriptor, acceptOp) :
    TryCompleteConnect(_fileDescriptor, (ConnectOperation)op);
```

pgavlin replied:
It depends on the size of the {true,false}-expressions. Seems okay in this case.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Linux.cs:
```diff
@@ -0,0 +1,139 @@
+                            var engine = new SocketAsyncEngine(epollFd);
+                            Task.Factory.StartNew(o => {
+                                ((SocketAsyncEngine)o).EventLoop();
+                            }, engine, TaskCreationOptions.LongRunning);
```
I'd suggest explicitly adding TaskScheduler.Default here as the scheduler argument.  Two related reasons:
- You want to make sure this actually targets the thread pool scheduler even if it happens to be called from a task running on another scheduler.
- The behavior of LongRunning is controlled by the scheduler: the thread pool scheduler gives it a dedicated thread, but another scheduler could behave differently.

pgavlin replied:
FWIW, I only used `Task` here b/c System.Net.Sockets did not have a reference to the System.Threading.Thread contract and I wasn't sure if it was okay to add one. If it is, I'd rather just use `Thread`.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Linux.cs:
```diff
@@ -0,0 +1,139 @@
+                {
+                    var handle = (GCHandle)events[i].data;
+                    var context = (SocketAsyncContext)handle.Target;
+                    context.HandleEvents(events[i].events);
```
There are a variety of places we throw exceptions from HandleEvents.  It seems like that'll end up preventing any further socket activity anywhere in the system. Should we be catching such exceptions and recovering in some way?

pgavlin replied:
Frankly, we probably shouldn't be throwing these exceptions at all: they're mostly a debugging aid. The error handling in general needs to be cleaned up.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Linux.cs:
```diff
@@ -0,0 +1,139 @@
+            var events = stackalloc Interop.libc.epoll_event[64];
+            for (;;)
+            {
+                int numEvents = Interop.libc.epoll_wait(_epollFd, events, 64, -1);
```
As I noted when we spoke yesterday, we should confirm what the behavior is of epoll_wait when there are no file descriptors left in its context.  If it blocks indefinitely, great.  If it immediately wakes up, this has become a busy loop.  something to check.

Additionally, I'm a bit concerned about this thread for the EventLoop sticking around forever, even when there's no work to be done.  It's probably not a big deal: after all, it's only a single thread, and it's rare for an application that uses Sockets to use them only briefly and then never again... more likely is that the application is using Sockets throughout the lifetime of the app.  Something to think about, though.

pgavlin replied:
I've confirmed that `epoll_wait` blocks indefinitely even when it is not tracking any file descriptors. Your point about the event loop thread sticking around is noted.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs:
```diff
@@ -0,0 +1,1098 @@
+            int optLevel, optName;
+            GetPlatformOptionInfo(optionLevel, optionName, out optLevel, out optName);
+
+            uint optLen = 4; // sizeof(int)
```
Just curious, why not `uint optLen = sizeof(int);` instead of `uint optLen = 4; // sizeof(int)` ?

pgavlin replied:
The original Windows code was written this way :)

stephentoub commented:
Excellent progress, Pat. Thanks.
